command_line_tools.md

```diff
@@ -0,0 +1,71 @@
+# Command Line Tools
+
+When writing command line tools, there are a couple of pitfalls to be aware of.
+In the past we made certain painful experiences which we would like to prevent
+in the future.
+
+### Project Structure
+
+Let's consider an example project for our new command line tool called
+`cooltool`. The project structure should look something like this.
+
+```
+$ tree cooltool -d
+cooltool
+└── cmd
+    ├── create
+    │   └── cluster
+    │       ├── defaulting
+    │       └── validation
```

Comment:
What is defaulting and validation? Do we have examples anywhere?

Reply:
* https://github.com/giantswarm/gsctl/tree/e2bb690400bc5aecc90a839e0f1e2c45ae36fb6b/cmd/cluster/scale/defaulting
* https://github.com/giantswarm/api/tree/034bdb4b5d7b1795b8f8ceeffdf4689ba09c051a/service/cluster/creator/defaulting
* https://github.com/giantswarm/api/tree/034bdb4b5d7b1795b8f8ceeffdf4689ba09c051a/service/cluster/creator/validation
* https://github.com/giantswarm/api/tree/034bdb4b5d7b1795b8f8ceeffdf4689ba09c051a/service/cluster/updater/defaulting
* https://github.com/giantswarm/api/tree/034bdb4b5d7b1795b8f8ceeffdf4689ba09c051a/service/cluster/updater/validation

Reply:
I don't understand the `Request` concept and how it's related to CLI commands. This comment about making that top level (https://github.com/giantswarm/sig-operator/pull/6/files#r254649442) confuses me even more. I think I'm not enough into details to share my opinion. I'll have to trust others on that but please not overdo something very simple. Also please not translate one language into another (`not` function is one clear sign of that).

Reply:
Tuomas suggested the `not` function for readability reasons and I think this was a good shout. You have a lot of these syntax sugar helpers in testing frameworks. I think that is legit. 

`Request` is maybe a wrong term. Here it gets a bit into implementation details. What we want is separation of concerns. We want to have isolated defaulting and validation so that we do not end up again in some unmanageable mess where EVERYTHING is in the same package and once you touch a line 16 different things break down. So we want to have cleaner APIs between packages. This is how we get the isolation. Then you can have the same structure everywhere and glue code to map types around accordingly, or, you have a types package you use within a specific domain. That types package here was called `Request`. We can call it `Input` or whatever. The name is not the important part but the idea behind it. So you have to die at least one death, because otherwise you get import cycles. When there are STILL issues or something unclear let's please get into a Meet and talk directly in person. This Github comment shit chat back and forth is not really productive IMO and we are usually better at getting to consensus when communicating directly.

Reply:
For the record: We had a chat with @xh3b4sd. While I understand the reasons of having defaulting/validations in common place _my_ preference would be to have that structure flattened. I.e. we would have files like `defaulting_worker.go`, `defaulting_scaling.go`, `defaulting_abc.go` or just `defaulting.go` if the `Request` is simple. Same applies for the validation. I understand there are drawbacks to this approach. Like we would have to have struct names like `defaulitngWorker` (but it's ok for _me_). It also would be difficult to introduce this to projects which aren't well structured right now.

I emphasized word "me" in the paragraph above because I think to a degree this is a matter of taste, personal workflows and tradeoffs we are willing to make. I think encapsulation is important and I don't want to leak some defaulting which is used only by one package. _At the moment_ I think it's the command itself that should be unit of encapsulation. In general I think it's easier to extract packages later as they are needed and that leads to a good organic design as opposed to package flattening which I think hardly ever happens.

 It would be cool to hear more voices but I guess it will be 4 ppl PR at the end of the day. That's why I'm going to approve this PR in its current shape. Still I wanted to share my thoughts. Also nothing is written in stone. Let's roll.

Reply:
And I think `Input` is a better name than `Request` when it comes to CLI commands :P